client/allocrunner/task_coordinator.go

```diff
@@ -0,0 +1,386 @@
+package allocrunner
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// taskCoordinatorState represents a state of the task coordinatator FSM.
+type taskCoordinatorState string
+
+const (
+	taskCoordinatorStateInit      taskCoordinatorState = "init"
+	taskCoordinatorStatePrestart                       = "prestart"
+	taskCoordinatorStateMain                           = "main"
+	taskCoordinatorStatePoststart                      = "poststart"
+	taskCoordinatorStatePoststop                       = "poststop"
+)
+
+// taskLifecycle represents a lifecycle configuration of interest for the task
+// coordination. Not all possible combinations of hook x sidecar are defined,
+// only the ones that are relevant for task coordinator.
+type taskLifecycle string
+
+const (
+	// taskLifecyclePrestartEphemeral are tasks with the "prestart" hook and
+	// sidecar set to "false".
+	taskLifecyclePrestartEphemeral taskLifecycle = "prestart_ephemeral"
+
+	// taskLifecyclePrestartSidecar are tasks with the "prestart" hook and
+	// sidecar set to "true".
+	taskLifecyclePrestartSidecar = "prestart_sidecar"
+
+	// taskLifecycleMain are tasks without a lifecycle or a lifecycle with an
+	// empty hook value.
+	taskLifecycleMain = "main"
+
+	// taskLifecyclePoststart are tasks with the "poststart" hook.
+	taskLifecyclePoststart = "poststart"
```

Comment:
I've been thinking about this, and initially I thought it wouldn't matter. 

The only way a task can run again is with a restart and currently the restart only affects task that are running. So in your example, the DB initialization task will always run only once, unless you happen to send a restart command while the task is still running.

The work I'm doing will allow for a restart command to affect all tasks, so the DB initialization task will run again, but in the expected order. Future work could allow restarting a `dead` task, so this would be relevante for that.

But then I thought about task recovery, and _maybe_ this could be a problem? I wasn't able to fully repro, but just in case this seems like the correct assertion. It also made me re-think how the next state is calculated, so it was worth another look 😄

Reply:
> The only way a task can run again is with a restart and currently the restart only affects task that are running. 

Ah right, that's a good point. I think there's probably a use case for allowing a task like this to be started again but that's totally out of scope for this PR.